File: erlang/seminars/transport/src/network.erl
-module(network).

-record(msg, {src=0, dst=0, data=[]}).

-export([start/2]).

start(Master, Id) ->
    Con = spawn(fun() -> init(Master, Id) end),
    {ok, Con}.

init(Master, Id) ->
    receive 
	{connect, Link} ->
	    io:format("network layer ~w connected~n", [Id]),	    
	    network(Master, Id, Link);
	quit ->
	    ok
    end.

network(Master, Id, Link) ->
    receive
	{send, To, Msg} ->
	    io:format("network layer ~w sending ~w~n", [Id, Msg]),
	    Link ! {send, #msg{src=Id, dst=To,  data=Msg}},
	    network(Master, Id, Link);

	#msg{dst=Id, data=Msg} ->
	    io:format("network layer ~w received ~w~n", [Id, Msg]),
	    Master ! Msg,
	    network(Master, Id, Link);

	#msg{}=_ ->
	    network(Master, Id, Link);

	{master, New} ->
	    network(New, Id, Link);	    

	status -> 
	    io:format("master: ~w id: ~w  link: ~w~n", [Master, Id, Link]),
	    network(Master, Id, Link);	    

	quit ->
	    ok
    end.

	    
			
    

File: exercises/transport/src/network.ex
defmodule Network do

  require Record

  Record.defrecord(:netw, src: 0, dst: 0, data: nil)

  
 def start(master, id) do
   con = spawn(fn() -> init(master, id) end)
   {:ok, con}
 end

 def init(master, id) do
   :io.format("network ~w: process ~w started ~n", [id, self()])
   receive do
     {:connect, link} ->
       :io.format("network ~w: connected to ~w~n", [id, link])
       network(master, id, link)
     :quit ->
       :ok
   end
 end

 def network(master, id, link) do
   receive do
     {:send, to, msg} ->
       ##:io.format("network ~w sending ~w to ~w\n", [id, msg, to])
       send(link, {:send, netw(src: id, dst: to,  data: msg)})
       network(master, id, link)

     netw(dst: ^id, data: msg) ->
       ##:io.puts("network ~w receiving ~w\n", [id, msg])
       send(master, msg)
       network(master, id, link)

     netw() ->
       network(master, id, link)

     {:master, new} ->
       network(new, id, link)

     :status -> 
       :io.format("network ~w: id ~w, master: ~w, link: ~w~n", [self(), id, master, link])
       network(master, id, link)

     :quit ->
       :ok
   end
 end

end	    
			
    







